src/components/annual-report/report-data/section4/useFetchSectionFourData.js

```diff
@@ -0,0 +1,111 @@
+import { useDataEngine } from '@dhis2/app-runtime'
+import { useState } from 'react'
+
+const section4ByLevel = {
+    data_detail_by_level: {
+        resource: 'analytics.json',
+        params: ({ dataElements, orgUnits, orgUnitLevel, periods }) => ({
+            dimension: `dx:${dataElements.join(';')},ou:${
+                orgUnits.join(';') + ';' + orgUnitLevel
+            },pe:${periods.join(';')}`,
+        }),
+    },
+}
+
+const section4Overall = {
+    data_over_all_org_units: {
+        resource: 'analytics.json',
+        params: ({ dataElements, orgUnits, periods }) => ({
+            dimension: `dx:${dataElements.join(';')},ou:${orgUnits.join(
+                ';'
+            )},pe:${periods.join(';')}`,
+        }),
+    },
+}
+
+export const fetchDataByLevel = async ({ engine, variables }) => {
+    const dataBySubPeriod = await engine.query(section4ByLevel, {
+        variables,
+    })
+    return dataBySubPeriod
+}
+
+export const useFetchSectionFourData = () => {
+    const [loading, setLoading] = useState(false)
+    const [data, setData] = useState(null)
+    const [error, setError] = useState(null)
+
+    const engine = useDataEngine()
+
+    const refetch = async ({ variables = {} }) => {
+        const { denominatorRelations } = variables?.mappedConfiguration
+        const denominatorRelationDEs = [
+            ...denominatorRelations?.map((rel) => rel.A.id),
+            ...denominatorRelations?.map((rel) => rel.B.id),
+        ]
+
+        // set to loading
+        setLoading(true)
+        // get data sets and check that each data element is associated with only one period type
+
+        // we need to preserve information on index to tie response back to relevant denominator relation (some can be filtered out)
+        const byLevelRequestsWithIndices = denominatorRelations
+            .map((rel, index) => {
+                // UN type does not require details by subOrgUnit level, so these can be skipped
+                if (rel.type === 'un') {
+                    return
+                }
+
+                // the level will be the minimum of the lowest of the denominator levels and the selected level
+                const denomMinLevel = Math.min(
+                    rel.A.lowLevel,
+                    rel.B.lowLevel,
+                    variables.orgUnitLevelNumber
+                )
+                // if the denominator relation level is less than the boundary org unit's level, it cannot be calculated
+                if (denomMinLevel < variables.boundaryOrgUnitLevel) {
+                    return
+                }
```

Comment:
I saw in the calculations doc "if either denominator is missing a DE/indicator: skip the denom. rel. calculation" -- is that handled here? (sorry if I'm missing it 🙃)

Reply:
Yeah. That should be handled by filtering out denominator relations without ids (or levels) when we summarize from the configuration: https://github.com/hisprwanda/who-data-quality-annual-report/blob/933933c10bfeb54ffc9d25213d56c908194fb47a/src/utils/utils.js#L98

Reply:
Great thanks! 🙏